itests/mempool_test.go

```diff
@@ -0,0 +1,455 @@
+//stm: #integration
+package itests
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/filecoin-project/go-state-types/big"
+	"github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/filecoin-project/lotus/itests/kit"
+	"github.com/stretchr/testify/require"
+)
+
+func TestMemPoolPushSingleNode(t *testing.T) {
+	//stm: @CHAIN_MEMPOOL_CREATE_MSG_CHAINS_001, @CHAIN_MEMPOOL_SELECT_001
+	//stm: @CHAIN_MEMPOOL_PENDING_001, @CHAIN_STATE_WAIT_MSG_001, @CHAIN_MEMPOOL_CAP_GAS_FEE_001
+	//stm: @CHAIN_MEMPOOL_PUSH_002
+	ctx := context.Background()
+	const blockTime = 100 * time.Millisecond
+	firstNode, _, _, ens := kit.EnsembleTwoOne(t, kit.MockProofs())
+	ens.InterconnectAll().BeginMining(blockTime)
+	kit.QuietMiningLogs()
+
+	sender := firstNode.DefaultKey.Address
+
+	addr, err := firstNode.WalletNew(ctx, types.KTBLS)
+	require.NoError(t, err)
+
+	const totalMessages = 10
+
+	bal, err := firstNode.WalletBalance(ctx, sender)
+	require.NoError(t, err)
+	toSend := big.Div(bal, big.NewInt(10))
+	each := big.Div(toSend, big.NewInt(totalMessages))
+
+	// add messages to be mined/published
+	var sms []*types.SignedMessage
+	for i := 0; i < totalMessages; i++ {
+		msg := &types.Message{
+			From:  sender,
+			To:    addr,
+			Value: each,
+		}
+
+		sm, err := firstNode.MpoolPushMessage(ctx, msg, nil)
+		require.NoError(t, err)
+		require.EqualValues(t, i, sm.Message.Nonce)
+
+		sms = append(sms, sm)
+	}
+
+	// check pending messages for address
+	msgStatuses, _ := firstNode.MpoolCheckPendingMessages(ctx, sender)
+	require.Equal(t, totalMessages, len(msgStatuses))
+	for _, msgStatusList := range msgStatuses {
+		for _, status := range msgStatusList {
+			require.True(t, status.OK)
+		}
+	}
+
+	// verify messages should be the ones included in the next block
+	selected, _ := firstNode.MpoolSelect(ctx, types.EmptyTSK, 0)
+	for _, msg := range sms {
+		found := false
+		for _, selectedMsg := range selected {
+			if selectedMsg.Cid() == msg.Cid() {
+				found = true
+				break
+			}
+		}
+		require.True(t, found)
+	}
+
+	time.Sleep(10 * blockTime)
+
+	// pool pending list should be empty
+	pending, err := firstNode.MpoolPending(context.TODO(), types.EmptyTSK)
+	require.NoError(t, err)
+	require.Equal(t, 0, len(pending))
+
+	// all messages should be added to the chain
+	for _, lookMsg := range sms {
+		msgLookup, err := firstNode.StateWaitMsg(ctx, lookMsg.Cid(), 3, api.LookbackNoLimit, true)
+		require.NoError(t, err)
+		require.NotNil(t, msgLookup)
+	}
+}
+
+func TestMemPoolPushTwoNodes(t *testing.T) {
+	//stm: @CHAIN_MEMPOOL_CREATE_MSG_CHAINS_001, @CHAIN_MEMPOOL_SELECT_001
+	//stm: @CHAIN_MEMPOOL_PENDING_001, @CHAIN_STATE_WAIT_MSG_001, @CHAIN_MEMPOOL_CAP_GAS_FEE_001
```

Comment:
I think grouping annotations on start of the function is not the best idea.  
While it works with STM, they would better show where the particular test case is in the test function  
if they were placed just above the said test case.

Reply:
I'll hold off on merge until you guys follow up on this

Reply:
@TheMenko Generally that's the best approach, but in this particular case those behaviors are happening "under the hood" (deeper in the call chain) so this is kinda the best way to write these annotations.